core/dbt/include/global_project/macros/materializations/models/table/columns_spec_ddl.sql

```diff
@@ -37,13 +37,32 @@
   {#--Obtain the column schema provided by the schema file by generating an 'empty schema' query from the model's columns. #}
   {%- set schema_file_provided_columns = get_column_schema_from_query(get_empty_schema_sql(model['columns'])) -%}
 
-  {%- set sql_file_provided_columns_formatted = format_columns(sql_file_provided_columns)  -%}
-  {%- set schema_file_provided_columns_formatted = format_columns(schema_file_provided_columns)  -%}
+  {#-- For compiler error msg #}
+  {%- set sql_columns = (format_columns(sql_file_provided_columns)|trim)  -%}
+  {%- set yaml_columns = (format_columns(schema_file_provided_columns)|trim)  -%}
 
-  {%- if sql_file_provided_columns_formatted != schema_file_provided_columns_formatted -%}
-    {%- do exceptions.raise_compiler_error('Please ensure the name, data_type, order, and number of columns in your `yml` file match the columns in your SQL file.\nSchema File Columns: ' ~ (schema_file_provided_columns_formatted|trim) ~ '\n\nSQL File Columns: ' ~ (sql_file_provided_columns_formatted|trim) ~ ' ' ) %}
+  {%- if sql_file_provided_columns|length != schema_file_provided_columns|length -%}
+    {%- do exceptions.raise_contract_error(yaml_columns, sql_columns) -%}
   {%- endif -%}
 
+  {%- for sql_col in sql_file_provided_columns -%}
+    {%- set yaml_col = [] -%}
+    {%- for schema_col in schema_file_provided_columns -%}
+      {%- if schema_col.name == sql_col.name -%}
+        {%- do yaml_col.append(schema_col) -%}
+        {%- break -%}
+      {%- endif -%}
+    {%- endfor -%}
+    {%- if not yaml_col -%}
+      {#-- Column with name not found in yaml --#}
+      {%- do exceptions.raise_contract_error(yaml_columns, sql_columns) -%}
+    {%- endif -%}
+    {%- if sql_col.dtype != yaml_col[0].dtype -%}
```

Comment:
Previously column types were compared based on their formatted representation, which could be data platform specific as implemented by the `<adapter>__format_column` macro. 

For example, [BigQuery's `format_column` implementation](https://github.com/dbt-labs/dbt-bigquery/pull/551/files#diff-ba5d875a4144a1b739519c0133ae7847a58715ca5374e841c06df089894778faR19) compares data_type values by `column.data_type` as opposed to `column.dtype` in order to make comparisons on nested data types. 

Strictly comparing SQL and yml data_type values by `dtype` would allow the following contract to be accepted in BigQuery:

```sql
SELECT
STRUCT("test" AS name, [1.0,2.0] AS laps) as some_struct
```

```yaml
models:
  - name: test_schema
    config:
      contract: true
    columns:
      - name: some_struct
        data_type: STRUCT<name FLOAT64, laps STRING> #wrong! but accepted because dtype == STRUCT for both SQL and schema.yml
```

One workaround would be to do this comparison using `format_column` instead, i.e: `adapter.dispatch('format_column', 'dbt')(sql_col) != adapter.dispatch('format_column', 'dbt')(yaml_col[0])`. This would also ensure the comparison and error messaging are using consistent logic. An alternative would be to implement a `default__compare_data_types` macro to enable adapter-specific implementations.

Reply:
I looked at using format_column but the default implementation of that just uses the dtype. Is that different in the other adapters?

Reply:
I don't actually see any implementations of format_column in the adapters.

Reply:
Oh, I was on the wrong bigquery branch. Bigquery is the only adapter that implements it.

Reply:
Could we include the formatted column in the Column structures returned by "get_column_schema_from_query"?

Reply:
Or concatenate the other parts of the column structure for comparison?